File: tests/rosetta/transpiler/Pascal/bitmap-ppm-conversion-through-a-pipe.pas
{$mode objfpc}
program Main;
uses SysUtils;
type Pixel = record
  R: integer;
  G: integer;
  B: integer;
end;
type PixelArray = array of Pixel;
type PixelArrayArray = array of PixelArray;
type Bitmap = record
  cols: integer;
  rows: integer;
  px: array of PixelArray;
end;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  y: integer;
  p: Pixel;
  x: integer;
  seed: integer;
function makeBitmap(cols: integer; rows: integer; px: PixelArrayArray): Bitmap; forward;
function makePixel(R: integer; G: integer; B: integer): Pixel; forward;
function pixelFromRgb(c: integer): Pixel; forward;
function rgbFromPixel(p: Pixel): integer; forward;
function NewBitmap(x: integer; y: integer): Bitmap; forward;
procedure FillRgb(b: Bitmap; c: integer); forward;
function SetPxRgb(b: Bitmap; x: integer; y: integer; c: integer): boolean; forward;
function nextRand(seed: integer): integer; forward;
procedure main(); forward;
function makeBitmap(cols: integer; rows: integer; px: PixelArrayArray): Bitmap;
begin
  Result.cols := cols;
  Result.rows := rows;
  Result.px := px;
end;
function makePixel(R: integer; G: integer; B: integer): Pixel;
begin
  Result.R := R;
  Result.G := G;
  Result.B := B;
end;
function pixelFromRgb(c: integer): Pixel;
var
  pixelFromRgb_r: integer;
  pixelFromRgb_g: integer;
  pixelFromRgb_b: integer;
begin
  pixelFromRgb_r := Trunc(c div 65536) mod 256;
  pixelFromRgb_g := Trunc(c div 256) mod 256;
  pixelFromRgb_b := c mod 256;
  exit(makePixel(pixelFromRgb_r, pixelFromRgb_g, pixelFromRgb_b));
end;
function rgbFromPixel(p: Pixel): integer;
begin
  exit(((p.R * 65536) + (p.G * 256)) + p.B);
end;
function NewBitmap(x: integer; y: integer): Bitmap;
var
  NewBitmap_data: array of PixelArray;
  NewBitmap_row: integer;
  NewBitmap_r: array of Pixel;
  NewBitmap_col: integer;
begin
  NewBitmap_data := [];
  NewBitmap_row := 0;
  while NewBitmap_row < y do begin
  NewBitmap_r := [];
  NewBitmap_col := 0;
  while NewBitmap_col < x do begin
  NewBitmap_r := concat(NewBitmap_r, [makePixel(0, 0, 0)]);
  NewBitmap_col := NewBitmap_col + 1;
end;
  NewBitmap_data := concat(NewBitmap_data, [NewBitmap_r]);
  NewBitmap_row := NewBitmap_row + 1;
end;
  exit(makeBitmap(x, y, NewBitmap_data));
end;
procedure FillRgb(b: Bitmap; c: integer);
var
  FillRgb_px: array of PixelArray;
  FillRgb_row: array of Pixel;
begin
  y := 0;
  p := pixelFromRgb(c);
  while y < b.rows do begin
  x := 0;
  while x < b.cols do begin
  FillRgb_px := b.px;
  FillRgb_row := FillRgb_px[y];
  FillRgb_row[x] := p;
  FillRgb_px[y] := FillRgb_row;
  b.px := FillRgb_px;
  x := x + 1;
end;
  y := y + 1;
end;
end;
function SetPxRgb(b: Bitmap; x: integer; y: integer; c: integer): boolean;
var
  SetPxRgb_px: array of PixelArray;
  SetPxRgb_row: array of Pixel;
begin
  if (((x < 0) or (x >= b.cols)) or (y < 0)) or (y >= b.rows) then begin
  exit(false);
end;
  SetPxRgb_px := b.px;
  SetPxRgb_row := SetPxRgb_px[y];
  SetPxRgb_row[x] := pixelFromRgb(c);
  SetPxRgb_px[y] := SetPxRgb_row;
  b.px := SetPxRgb_px;
  exit(true);
end;
function nextRand(seed: integer): integer;
begin
  exit(((seed * 1664525) + 1013904223) mod 2147483648);
end;
procedure main();
var
  main_bm: Bitmap;
  main_i: integer;
begin
  main_bm := NewBitmap(400, 300);
  FillRgb(main_bm, 12615744);
  seed := _now();
  main_i := 0;
  while main_i < 2000 do begin
  seed := nextRand(seed);
  x := seed mod 400;
  seed := nextRand(seed);
  y := seed mod 300;
  SetPxRgb(main_bm, x, y, 8405024);
  main_i := main_i + 1;
end;
  x := 0;
  while x < 400 do begin
  y := 240;
  while y < 245 do begin
  SetPxRgb(main_bm, x, y, 8405024);
  y := y + 1;
end;
  y := 260;
  while y < 265 do begin
  SetPxRgb(main_bm, x, y, 8405024);
  y := y + 1;
end;
  x := x + 1;
end;
  y := 0;
  while y < 300 do begin
  x := 80;
  while x < 85 do begin
  SetPxRgb(main_bm, x, y, 8405024);
  x := x + 1;
end;
  x := 95;
  while x < 100 do begin
  SetPxRgb(main_bm, x, y, 8405024);
  x := x + 1;
end;
  y := y + 1;
end;
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  main();
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/rosetta/transpiler/FS/bitmap-ppm-conversion-through-a-pipe.fs
// Generated 2025-07-26 10:43 +0700

exception Return

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
type Pixel = {
    R: int
    G: int
    B: int
}
type Bitmap = {
    cols: int
    rows: int
    px: Pixel array array
}
open System

let rec pixelFromRgb (c: int) =
    let mutable __ret : Pixel = Unchecked.defaultof<Pixel>
    let mutable c = c
    try
        let r: int = (((int (c / 65536)) % 256 + 256) % 256)
        let g: int = (((int (c / 256)) % 256 + 256) % 256)
        let b: int = ((c % 256 + 256) % 256)
        __ret <- unbox<Pixel> { R = r; G = g; B = b }
        raise Return
        __ret
    with
        | Return -> __ret
and rgbFromPixel (p: Pixel) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable p = p
    try
        __ret <- (((p.R) * 65536) + ((p.G) * 256)) + (p.B)
        raise Return
        __ret
    with
        | Return -> __ret
and NewBitmap (x: int) (y: int) =
    let mutable __ret : Bitmap = Unchecked.defaultof<Bitmap>
    let mutable x = x
    let mutable y = y
    try
        let mutable data: Pixel array array = [||]
        let mutable row: int = 0
        while row < y do
            let mutable r: Pixel array = [||]
            let mutable col: int = 0
            while col < x do
                r <- unbox<Pixel array> (Array.append r [|{ R = 0; G = 0; B = 0 }|])
                col <- col + 1
            data <- unbox<Pixel array array> (Array.append data [|r|])
            row <- row + 1
        __ret <- unbox<Bitmap> { cols = x; rows = y; px = data }
        raise Return
        __ret
    with
        | Return -> __ret
and FillRgb (b: Bitmap) (c: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable b = b
    let mutable c = c
    try
        let mutable y: int = 0
        let p: Pixel = pixelFromRgb c
        while y < (b.rows) do
            let mutable x: int = 0
            while x < (b.cols) do
                let mutable px: Pixel array array = b.px
                let mutable row: Pixel array = px.[y]
                row.[x] <- p
                px.[y] <- row
                b <- { b with px = px }
                x <- x + 1
            y <- y + 1
        __ret
    with
        | Return -> __ret
and SetPxRgb (b: Bitmap) (x: int) (y: int) (c: int) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable b = b
    let mutable x = x
    let mutable y = y
    let mutable c = c
    try
        if (((x < 0) || (x >= (b.cols))) || (y < 0)) || (y >= (b.rows)) then
            __ret <- false
            raise Return
        let mutable px: Pixel array array = b.px
        let mutable row: Pixel array = px.[y]
        row.[x] <- pixelFromRgb c
        px.[y] <- row
        b <- { b with px = px }
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
and nextRand (seed: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable seed = seed
    try
        __ret <- int ((((int64 ((seed * 1664525) + 1013904223)) % 2147483648L + 2147483648L) % 2147483648L))
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let mutable bm: Bitmap = NewBitmap 400 300
        FillRgb bm 12615744
        let mutable seed: int = _now()
        let mutable i: int = 0
        while i < 2000 do
            seed <- int (nextRand seed)
            let x: int = ((seed % 400 + 400) % 400)
            seed <- int (nextRand seed)
            let y: int = ((seed % 300 + 300) % 300)
            SetPxRgb bm x y 8405024
            i <- i + 1
        let mutable x: int = 0
        while x < 400 do
            let mutable y: int = 240
            while y < 245 do
                SetPxRgb bm x y 8405024
                y <- y + 1
            y <- 260
            while y < 265 do
                SetPxRgb bm x y 8405024
                y <- y + 1
            x <- x + 1
        let mutable y: int = 0
        while y < 300 do
            let mutable x: int = 80
            while x < 85 do
                SetPxRgb bm x y 8405024
                x <- x + 1
            x <- 95
            while x < 100 do
                SetPxRgb bm x y 8405024
                x <- x + 1
            y <- y + 1
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
